LetsDoIt.CustomValueTypes/src/LetsDoIt.CustomValueTypes/Email.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System;
+using System.Net.Mail;
+using System.Text.Json.Serialization;
+
+namespace LetsDoIt.CustomValueTypes
+{
+
+    [JsonConverter(typeof(EmailJsonConverter))]
+    public readonly struct Email
+    {
+        private readonly string _value;
+
+        public Email(string value)
+        {
+            if (value == null)
+            {
+                throw new ArgumentNullException(nameof(value), "Email can not be null!");
+            }
+
+            if (!IsValid(value))
+            {
+                throw new ArgumentException("Invalid email address.", value);
+            }
+
+            _value = value.ToLowerInvariant();
+        }
+
+        public static bool IsValid(string email)
+        {
+            if (string.IsNullOrEmpty(email))
+            {
+                return false;
+            }
+
+            try
+            {
+                var mailAddress = new MailAddress(email);
+
+                return string.Equals(mailAddress.Address, email, StringComparison.InvariantCultureIgnoreCase);
+            }
+            catch
+            {
+                return false;
+            }
+
+        }
+
+        public static bool TryParse(string candidate, out Email? email)
+        {
+            email = null;
+
+            if (string.IsNullOrWhiteSpace(candidate))
+            {
+                return false;
+            }
+
+            email = new Email(candidate);
+
+            return true;
+        }
+
+        public static Email Parse(string candidate)
+        {
+            if (string.IsNullOrWhiteSpace(candidate))
+            {
+                throw new ArgumentException("Email can not be empty!");
+            }
+
+            return new Email(candidate);
+        }
+
+        public static implicit operator string(Email email)
+        {
+            return email._value;
+        }
+
+        public static implicit operator Email(string email)
+        {
+            return new Email(email);
+        }
+
+        public override string ToString()
+        {
+            return _value;
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj is Email objEmail)
```

Comment:
if you can compare hashCodes instead of string equals here that method would work faster!

Reply:
Don't know how to do it

Reply:
`return GetHashCode() == objEmail.GetHashCode();`

Reply:
- _if two things are equal (Equals(...) == true) then they must return the same value for GetHashCode()_

- _if the GetHashCode() is equal, it is not necessary for them to be the same; this is a collision, and Equals will be called to see if it is a real equality or not._

### We still need equals, I guess. I am not sure of the correct way of equalizing.

https://stackoverflow.com/questions/371328/why-is-it-important-to-override-gethashcode-when-equals-method-is-overridden

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/how-to-define-value-equality-for-a-type

Reply:
Yes, That makes sense. We can keep it as is